File: apps/windows/Finance/FinanceJson.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Andivum_Windows.Finance;

public static class FinanceJson
{
    private static readonly JsonSerializerOptions Options = new(JsonSerializerDefaults.Web)
    {
        WriteIndented = false,
    };

    public static string CreateTransactionPayload(
        FinanceDraft draft,
        string accountId,
        string source = "manual",
        string? importFingerprint = null)
    {
        if (string.IsNullOrWhiteSpace(accountId))
        {
            throw new ArgumentException("Account is required.", nameof(accountId));
        }
        if (string.IsNullOrWhiteSpace(draft.Title))
        {
            throw new ArgumentException("Transaction title is required.");
        }

        if (draft.TotalMinor <= 0 || draft.Items.Count == 0)
        {
            throw new ArgumentException(
                "Transaction must have a positive total and at least one item.",
                nameof(draft));
        }

        if (draft.Items.Sum(item => item.LineTotalMinor) != draft.TotalMinor)
        {
            throw new ArgumentException("Items must equal the transaction total.");
        }

        var payload = new FinanceCreateTransactionPayload(
            accountId,
            draft.Type.ToString().ToLowerInvariant(),
            draft.Title.Trim(),
            draft.OccurredOn,
            draft.Currency.ToUpperInvariant(),
            draft.TotalMinor,
            source,
            importFingerprint,
            draft.Items.Select(item => new FinanceCreateItemPayload(
                item.Name.Trim(),
                item.Quantity,
                item.UnitPriceMinor,
                item.LineTotalMinor,
                item.CategorySlug)).ToArray());

        return JsonSerializer.Serialize(payload, Options);
    }

    private sealed record FinanceCreateTransactionPayload(
        [property: JsonPropertyName("account_id")] string AccountId,
        [property: JsonPropertyName("type")] string Type,
        [property: JsonPropertyName("title")] string Title,
        [property: JsonPropertyName("occurred_on")] string OccurredOn,
        [property: JsonPropertyName("currency")] string Currency,
        [property: JsonPropertyName("total_minor")] long TotalMinor,
        [property: JsonPropertyName("source")] string Source,
        [property: JsonPropertyName("import_fingerprint")] string? ImportFingerprint,
        [property: JsonPropertyName("items")] IReadOnlyList<FinanceCreateItemPayload> Items);

    private sealed record FinanceCreateItemPayload(
        [property: JsonPropertyName("name")] string Name,
        [property: JsonPropertyName("quantity")] decimal Quantity,
        [property: JsonPropertyName("unit_price_minor")] long UnitPriceMinor,
        [property: JsonPropertyName("line_total_minor")] long LineTotalMinor,
        [property: JsonPropertyName("category_slug")] string CategorySlug);
}

File: apps/android/app/src/main/java/io/github/andivalsei/andivum/FinanceJson.kt
package io.github.andivalsei.andivum

import java.math.BigDecimal
import java.math.RoundingMode
import org.json.JSONArray
import org.json.JSONObject

object FinanceJson {
    fun createTransactionPayload(
        draft: FinanceDraft,
        accountId: String,
        source: String = "manual",
        importFingerprint: String? = null,
    ): JSONObject {
        require(accountId.isNotBlank()) { "Account is required." }
        require(draft.title.isNotBlank()) { "Transaction title is required." }
        require(draft.totalMinor > 0L && draft.items.isNotEmpty()) {
            "Transaction must have a positive total and at least one item."
        }
        require(draft.items.sumOf { it.lineTotalMinor } == draft.totalMinor) {
            "Items must equal the transaction total."
        }

        val items = JSONArray()
        draft.items.forEach { item ->
            require(item.name.isNotBlank()) { "Transaction item name is required." }
            require(item.quantity > 0.0 && item.lineTotalMinor >= 0L) {
                "Transaction item amount is invalid."
            }
            items.put(
                JSONObject()
                    .put("name", item.name.trim())
                    .put("quantity", item.quantity)
                    .put("unit_price_minor", item.unitPriceMinor)
                    .put("line_total_minor", item.lineTotalMinor)
                    .put("category_slug", item.categorySlug),
            )
        }

        return JSONObject()
            .put("account_id", accountId)
            .put("type", draft.type.name.lowercase())
            .put("title", draft.title.trim())
            .put("occurred_on", draft.occurredOn)
            .put("currency", draft.currency.uppercase())
            .put("total_minor", draft.totalMinor)
            .put("source", source)
            .putOpt("import_fingerprint", importFingerprint)
            .put("items", items)
    }
}

object FinanceMoney {
    fun parseMinorUnits(value: String, currency: String): Long {
        val normalized = value.trim()
            .replace(" ", "")
            .replace(',', '.')
        require(normalized.isNotBlank()) { "Amount is required." }
        val scale = when (currency.uppercase()) {
            "BHD", "IQD", "JOD", "KWD", "LYD", "OMR", "TND" -> 3
            "CLP", "DJF", "GNF", "ISK", "JPY", "KMF", "KRW", "PYG", "RWF", "UGX", "VND", "VUV", "XAF", "XOF", "XPF" -> 0
            else -> 2
        }
        return BigDecimal(normalized)
            .movePointRight(scale)
            .setScale(0, RoundingMode.HALF_UP)
            .longValueExact()
    }
}
